### PR TITLE
Propagate target platform to builds

### PR DESCRIPTION
**Pull request by robertgzr**
This ensures that the build resolves the correct platform for base
images it fetches from the registry.
Includes unit tests to check the platform translation and the actual
build execution.

NOTE: This requires running a docker daemon >=18.09.3 and experimental
features enabled.

Change-type: minor
Signed-off-by: Robert Günzler <robertg@balena.io>

### CODE REVIEW

**Comment by robertgzr** [comment]
@kaisoz I managed to write all the tests I needed right here. I haven't tried to pull this into the builder itself yet, but the behavior of this module should be correct now.

**Comment by robertgzr** [comment]
@balena-ci retest

**Comment by robertgzr** [comment]
looks like circle does not support enabling experimental docker features using the `setup_remote_docker` builtin: https://ideas.circleci.com/ideas/CCI-I-491

**Comment by robertgzr** [comment]
@CameronDiver any idea how to approach the circleci failures?

**Comment by robertgzr** [comment]
@kaisoz @CameronDiver can you have another look please. I made the new test optional skip if experimental features are not enabled on the daemon we test against.

I also changed the resolve code to not transform the `dockerOpts` at all if there's no matching platform.

let me know if this looks good and I will rebase

File: lib/resolve.ts [hunk 1]
```diff
@@ -45,6 +45,13 @@ export function resolveTask(
 	additionalVars: Dictionary<string> = {},
 	dockerfilePreprocessHook?: (content: string) => string,
 ): BuildTask {
+	// do this first, to also signal proper platform to pulls
+	// and fall back to old behavior, if no match was found
+	const platform = resolveDockerPlatform(architecture);
+	if (platform) {
+		task.dockerOpts = _.merge({ platform }, task.dockerOpts);
```
Comment (CameronDiver) [inline lib/resolve.ts:52]:
```suggestion
		task.dockerOpts = { platform, task.dockerOpts };
```

Comment (robertgzr) [reply lib/resolve.ts:52]:
```suggestion
		task.dockerOpts = { platform, ...task.dockerOpts };
```

works

**Comment by robertgzr** [comment]
@balena-ci retest

**Comment by robertgzr** [comment]
All changes squashed into a single commit now. will merge if the tests pass

**Pull request merged**